hell/adventcalendar/2025/3/index.md

```diff
@@ -1,35 +1,123 @@
+
+Using mouse hover as a signal works well on desktop, but what about mobile, where "hover" is not really a thing? At least in browser terms — I'm sure many of us have hovered over a link without fingers, but as that is not detected by the browser as an event, we can't act upon it.
+
+Well, solving that is just one of many improvements the API has seen in the last year that we'll cover in this post.
+
+## Improved mobile viewport heuristics
+
+Since hover does not exist, we have updated the `eagerness: moderate` on mobile to instead look at the following heuristics:
+
+- Anchor links within 30% vertical distance from the previous pointer down.
+- Anchor links at least 0.5× as big as the largest anchor in the viewport.
+- The browser waits 500 milliseconds after the user stopped scrolling.
+
+The aim is to avoid, as much as possible, over-speculating links that users are less likely to click. For example, because they are still scrolling, or they are very small links (terms and conditions and the like!).
+
+You can see this in action in the following video where the [Web Almanac](https://almanac.httparchiv.org) prerenders same-origin links with `moderate` eagerness:
+
+<figure>
+  <video controls playsinline muted loop>
+      <source src="/images/advent2025/speculation-rules-mobile-viewport-demo.mp4" width=2420 height=1320>
+  </video>
+  <figcaption><p>An example of scrolling down a mobile page triggering speculations.</p></figcaption>
+</figure>
+<script>
+  const mediaQuery = window.matchMedia('(prefers-reduced-motion: reduce)');
+  if (!mediaQuery.matches) {
+      document.querySelector('video').autoplay = true;
+  }
+</script>
+
+In the video, you can see that as you scroll down the page, links are being successfully prerendered, ready for the user to browse to the page.
+
+<p class="highlight"><strong>Note:</strong> Speculating has a cost. To both your users and for sites with potential increased traffic and resulting infrastructure usage. Always weigh those costs against the benefit to the user.</p>
+
+To conserve memory, Chrome keeps up to two speculations in memory at a time. As the user scrolls further and new links — which are more likely to be clicked on — enter the viewport, the old prerenders are cancelled. Should these links be re-speculated (for example, but the user scrolls back up), then they will be fetched from the HTTP cache and so prerender even faster.
 
-## Heading
+## `eager` eagerness improvements
 
-Some text.
+As well as the improvements to the `moderate` eagerness, the `eager` value has also been recently changed to offer an option somewhere between `immediate` (where links are speculated as soon as possible) and `moderate`.
 
+On desktop, `moderate` rules trigger after a 10 millisecond hover, and on mobile we consider ALL links in the viewport after the user has stopped scrolling for 100 milliseconds, rather than the more restrictive set of heuristics above.
```

Comment:
for readability: 

```suggestion
On desktop, `moderate` rules trigger after a 10 millisecond hover. On mobile, we consider _all_ links in the viewport after the user has stopped scrolling for 100 milliseconds, rather than the more restrictive set of heuristics above.
```